File: lib/futures/books/domain/entites/book.dart
import 'package:equatable/equatable.dart';

abstract class Book extends Equatable {
  final int id;
  final String title;
  final String content;
  final String author;
  final dynamic rate;
  final int totalpages;
  final String img;
  final String audio;
  final List<dynamic> tags;
  final List<dynamic> catigories;
  final String file;

  const Book(
      {required this.id,
      required this.title,
      required this.author,
      required this.content,
      required this.rate,
      required this.totalpages,
      required this.img,
      required this.audio,
      required this.tags,
      required this.catigories,
      required this.file});

  @override
  List<Object?> get props =>
      [title, content, author, rate, img, audio, file, totalpages];
}

File: app/Models/Book.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use App\Traits\Searchable;
use App\Utilities\FilterBuilder;

class Book extends Model
{
    use HasFactory;
    use Searchable;

    protected $fillable = [
        'title',
        'content',
        'author',
        'rate',
        'totalpages',
        'img',
        'audio',
        'tags',
        'file',
        'category_id'
    ];

    public function category()
    {
        return $this->belongsTo(Category::class);
    }

    public function scopeSearch($query, $value)
    {
        $columns = ['title', 'author', 'content'];
        foreach($columns as $column){
        $query->orWhere($column, 'LIKE', '%' . $value->value . '%')
              ->orWhereHas('category', function($q) use ($value, $column)
        {
            $q->where($column, 'LIKE', '%'.$value->value.'%');
        });   
        }
    }

    public function scopeFilterBy($query, $filters)
    {
        $namespace = 'App\Utilities\Filters';
        $filter = new FilterBuilder($query, $filters, $namespace);
        return $filter->apply();
    }
}
